retrofit/src/main/java/retrofit/Retrofit.java

```diff
@@ -108,35 +108,29 @@ private Retrofit(OkHttpClient client, Endpoint endpoint, Converter.Factory conve
   }
 
   private final InvocationHandler handler = new InvocationHandler() {
-    @Override public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+    @Override public Object invoke(Object proxy, Method method, Object... args) throws Throwable {
       // If the method is a method from Object then defer to normal invocation.
       if (method.getDeclaringClass() == Object.class) {
         return method.invoke(this, args);
       }
-      return invokeMethod(method, args);
+      return loadMethodInvoker(method).invoke(args);
     }
   };
 
-  // Package-private avoids synthetic accessor method from InvocationHandler. Also for testing.
-  Object invokeMethod(Method method, Object... args) {
-    MethodInfo methodInfo = loadMethodInfo(method);
-    Converter<?> responseConverter = methodInfo.responseConverter;
-    Call call = new OkHttpCall<>(client, endpoint, responseConverter, methodInfo, args);
-    return methodInfo.adapter.adapt(call);
-  }
-
-  private MethodInfo loadMethodInfo(Method method) {
-    MethodInfo methodInfo = methodInfoCache.get(method);
-    if (methodInfo == null) {
-      synchronized (methodInfoCache) {
-        methodInfo = methodInfoCache.get(method);
-        if (methodInfo == null) {
-          methodInfo = new MethodInfo(method, adapterFactory, converterFactory);
-          methodInfoCache.put(method, methodInfo);
+  // Package-private avoids synthetic accessor method from InvocationHandler.
+  MethodInvoker<?> loadMethodInvoker(Method method) {
+    MethodInvoker<?> methodInvoker = methodInvokerCache.get(method);
```

Comment:
Just use `synchronized`. It's fast unless there's contention, and there isn't going to be contention.